src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1450,4 +1450,76 @@ impl Cage {
             }
         }
     }
+    
+    //------------------GETDENTS SYSCALL------------------
+
+    pub fn getdents_syscall(&self, fd: i32, bufsize: usize, vec: &mut Vec<(ClippedDirent, Vec<u8>)>) -> i32 {
+        
+        // make sure bufsize is at least greater than size of a ClippedDirent struct
+        let dirent_size = std::mem::size_of::<ClippedDirent>();
+        if bufsize <= dirent_size {
+            return syscall_error(Errno::EINVAL, "getdents", "Buffer size too small");
+        }
+        
+        let fdtable = self.filedescriptortable.read().unwrap();
+
+        if let Some(wrappedfd) = fdtable.get(&fd) { // check if fd is valid
+            let mut filedesc_enum = wrappedfd.write().unwrap();
+            
+            match &mut *filedesc_enum {
+                // only proceed when fd represents a file
+                File(ref mut normalfile_filedesc_obj) => {
+                    let metadata = FS_METADATA.read().unwrap();
+                    let inodeobj = metadata.inodetable.get(&normalfile_filedesc_obj.inode).unwrap();
+
+                    match inodeobj {
+                        // only proceed when inode is a dir
+                        Inode::Dir(dir_inode_obj) => {
+                            let position = normalfile_filedesc_obj.position;
+                            let mut bufcount = 0;
+                            let mut curr_size = 0;
+                            let mut curr_pos = 0;
+                            let mut vec_idx = 0;
+                            let mut count = 0;
+
+                            for (filename, inode) in dir_inode_obj.filename_to_inode_dict.clone().into_iter() {
+                                if curr_pos >= position {
+                                    vec[vec_idx].1 = filename.as_bytes().to_vec();
+                                    for n in 0..(vec.len() + 7) / 8 * 8 - vec.len() {
+                                        vec[vec_idx].1.push(00);
+                                    }
+                                    vec[vec_idx].1.push(libc::DT_UNKNOWN);
```

Comment:
Also, it's supposed to be in place of the last padding byte, not in addition to it (as that would ruin alignment)

Reply:
So in order to make that work, you need to make sure that it pads with at leats 1 byte (which means in the case where vec.len is precisely divisible by 8, you must pad with 8 bytes--so here you should add 8 rather than to vec.len().
Then you need to subtract on to the end of that big term that is the number of iterations you push in so that the DT_UNKNOWN is in place of a padding byte rather than in addition to it.

Reply:
Rereading the spec, DT_UNKNOWN actually goes before the first byte that's strictly a padding byte, so the implementation of that should be relatively easy from your current implementation, just put it before your padding.
Additionally, d_reclen should be taken from this value rather than from the padded value.